### PR TITLE
implement amm deposits/withdrawals

### PR DESCRIPTION
**Pull request by adrian-gierakowski**
In order to deposit/withdraw, user first needs to query the api to find out
current ration for given pool. This is done by sending `GET
/v1/trading/amm/fundingOrderData` with `pool`, `token` and `amount`. If the
`token` if one of the pool token, the response will contain AmmFundingOrder
data corresponding to a deposit of the given amount of token specified, and
a required amount of the other pool token. The result of `GET
fundingOrderData` call can be inspected or presented to the user for
approval and then can be passed to `POST /v1/trading/amm/fundingOrder` for
execution. Withdrawals are done the same way, however in the call to `GET
fundingOrderData` the token needs to be set to the LP token for given pool,
and the amount to the amount of LP token the user is going to return to the
pool. The response of `GET fundingOrderData` will contain information about
how much of each pool token the user will get in return.

NOTE: amm deposits/withdrawals are performed on L2 by executing 2 starkware
limit orders, one for each pool token. The LP token received
(or given away) by the user are divided equally between these 2 orders. The
orders need to be signed with users stark private key (which is done
automatically by `src/api/postAmmFundingOrder.js`, using
`dvfStarkProvider`, hence no additional method is needed to sign using
ledger).

### CODE REVIEW

File: src/api/postAmmFundingOrder.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+const validateData = validateWithJoi(schema)('INVALID_METHOD_ARGUMENT')({
+  context: 'postAmmFundingOrder'
+})
+
+const endpoint = '/v1/trading/amm/postAmmFundingOrder'
+
+module.exports = async (dvf, data, nonce, signature) => post(
+  dvf, endpoint, nonce, signature, R.compose(
+    dvf.stark.singAmmFundingOrder,
```
Comment (arijoon) [inline src/api/postAmmFundingOrder.js:48]: is this meant to be `sign` instead of `sing`?

File: src/lib/dvf/bindApi.js [hunk 2]
```diff
@@ -31,6 +31,7 @@ module.exports = () => {
     createOrderMessage: compose(require('../stark/createOrderMessage')),
     sign: compose(require('../stark/starkSign')),
     signAuth: compose(require('../stark/starkSignAuth')),
+    singAmmFundingOrder: require('../stark/singAmmFundingOrder')(dvf),
```
Comment (arijoon) [inline src/lib/dvf/bindApi.js:34]: same typo `sing` instead of `sign`

**Review by arijoon — APPROVED** [review]

File: src/lib/stark/ledger/createSignedOrder.js [hunk 3]
```diff
@@ -1,13 +1,20 @@
...
-module.exports = async (dvf, path, starkOrder) => {
+const getPublicKey = async (eth, starkPath) => {
+  const tempKey = (await eth.starkGetPublicKey(starkPath)).toString('hex')
+  return {
+    x: tempKey.substr(2, 64),
+    y: tempKey.substr(66)
+  }
+}
+
+module.exports = async (dvf, path, starkOrder, { dontGetPublicKey = false } = {}) => {
```
Comment (Neozaru) [inline src/lib/stark/ledger/createSignedOrder.js:17]: Do we want to positivize `dontGetPublicKey` to `getPublicKey` ?

File: src/lib/stark/singAmmFundingOrder.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  ...data,
+  starkPublicKey: await dvf.dvfStarkProvider.getPublicKey(),
+  // Need to do this in series since createSignedTransaction might call
+  // into ledger and these calls cannot be executed in parallel.
+  orders: await P.mapSeries(
+    data.orders,
+    async order => ({
+      ...order,
+      starkOrder: order.starkOrder.signature
+        // Only sign if there is no signature
```
Comment (Neozaru) [inline src/lib/stark/singAmmFundingOrder.js:14]: Can we elaborate in what contexts that could happen ?

File: src/api/postAmmFundingOrder.js (hunk 1 above)
Comment (adrian-gierakowski) [reply src/api/postAmmFundingOrder.js]: fixed in bba3f9f49865cfe56a72b8f516231b5433754c08

File: src/lib/stark/singAmmFundingOrder.js (hunk 4 above)
Comment (adrian-gierakowski) [reply src/lib/stark/singAmmFundingOrder.js]: if someone used their own code to create and sign the `starkOrder`  they we wouldn't want to overwrite this signature. I guess you might ask why would they call `signAmmFundingOrder`? They might not call it directly but it called [here](https://github.com/DeversiFi/dvf-client-js/blob/bba3f9f49865cfe56a72b8f516231b5433754c08/src/api/postAmmFundingOrder.js#L48). But I think I might refactor things a bit and move the conditional out of this function

Comment (adrian-gierakowski) [reply src/lib/stark/singAmmFundingOrder.js]: fixed in 430423a5ab0512d79315c8efac8b481fc67977cb

File: src/lib/stark/ledger/createSignedOrder.js (hunk 3 above)
Comment (adrian-gierakowski) [reply src/lib/stark/ledger/createSignedOrder.js]: fixed in 09978ab60df1bb0707e7f20ffbbdcff3d694d491

File: src/lib/dvf/bindApi.js (hunk 2 above)
Comment (adrian-gierakowski) [reply src/lib/dvf/bindApi.js]: fixed in bba3f9f49865cfe56a72b8f516231b5433754c08

**Review by Neozaru — APPROVED** [review]

**Comment by Neozaru** [comment]
Is that ready to be merged ? Or is it waiting for some other work ?

**Comment by plutoegg** [comment]
> Is that ready to be merged ? Or is it waiting for some other work ?

Looks ready to be merged to me

**Comment by adrian-gierakowski** [comment]
I’ll resolve conflicts and merge tomorrow

File: examples/27.ammDeposit.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  const token2 = 'USDT'
+  const depositETHResponse = await dvf.deposit(token1, 0.1, starkPrivKey)
+  const depositUSDTResponse = await dvf.deposit(token2, 1000, starkPrivKey)
+
+  if (process.env.WAIT_FOR_DEPOSIT_READY === 'true') {
+    await waitForDepositCreditedOnChain(dvf, depositETHResponse)
+    await waitForDepositCreditedOnChain(dvf, depositUSDTResponse)
+  }
+
+  const pool = `${token1}:${token2}`
```
Comment (gvko) [inline examples/27.ammDeposit.js:50]: The pools in the AMM actually don't have `:` delimiter. Is this delimiter removed somewhere down the line of handling the request?

**Review by gvko — CHANGES_REQUESTED** [review]

File: examples/27.ammDeposit.js (hunk 5 above)
Comment (adrian-gierakowski) [reply examples/27.ammDeposit.js:50]: I believe pool value here is incorrect, it should have no delimiter

Comment (adrian-gierakowski) [reply examples/27.ammDeposit.js]: fixed in e30bd3d44a58c8101120e5658a359b0fcc4ed973

**Pull request merged**